ibis/expr/rewrites.py

```diff
@@ -132,28 +119,65 @@ def backtrack(cls, value):
         ):
             yield value, distance
 
-    def dereference(self, value):
-        """Dereference a value to the target relations.
+    def _fill_substitution_mappings(self):
+        if self.subs is not None and self.ambigs is not None:
+            return
+
+        mapping = defaultdict(dict)
+
+        for rel in self.rels:
+            for field in rel.fields.values():
+                for val, distance in self.__class__.backtrack(field):
+                    mapping[val][field] = distance
+
+        subs, ambigs = {}, {}
+        for from_, to in mapping.items():
+            mindist = min(to.values())
+            minkeys = [k for k, v in to.items() if v == mindist]
```

Comment:
not a big point but would this maybe be easier to express/marginally more efficient as:

```
to_sorted_by_dist = list(sorted(to.items(), key=lambda to_item: to_item[0]))
minkeys = []
mindist = to_sorted_by_dist[0][1]
for k, v in to_sorted_by_dist:
    if v == min:
        minkeys.append(k)
    else:
        break
```
the loop below over minkeys to distinguish the subs/ambigs could also be incorporated

Reply:
apologies after thinking further sorting would be more expensive than 2 loops over the list, so please ignore above comment